File: sdk/python/examples/price_levels.py
"""pAMM price levels: Titan's pre-quoted prices for a range of trade sizes.

Grouped per pAMM and served from Titan's latest snapshot over HTTP (no node
RPC, no key), unlike `router.quote`, which runs an on-chain `eth_call`.

Install the SDK first, then run:

    pip install -e .
    python3 examples/price_levels.py

Set PRICE_LEVELS_URL to use a specific HTTP endpoint.
"""

from __future__ import annotations

import asyncio
import os

from propamm import PriceLevels, PriceLevelsRpcSource
from propamm.common.helpers import format_units, parse_address
from propamm.common.pamms import PAMMS
from propamm.common.tokens import USDC, USDT, WETH

# Known tokens, for readable amounts. Unknown tokens print raw (0 decimals).
_WBTC = parse_address("0x2260fac5e5542a773aa44fbcfedf7c193bc2c599")
_TOKENS = {USDC: ("USDC", 6), USDT: ("USDT", 6), WETH: ("WETH", 18), _WBTC: ("WBTC", 8)}
# Venue address -> readable pAMM name.
_VENUE_NAMES = {address: name for name, address in PAMMS.items()}


def _token_info(token: str) -> tuple[str, int]:
    return _TOKENS.get(token, (token, 0))


def _fmt(amount: int, token: str) -> str:
    symbol, decimals = _token_info(token)
    return f"{format_units(amount, decimals)} {symbol}"


def _venue_name(address: str) -> str:
    return _VENUE_NAMES.get(address, address)


async def main() -> None:
    # PriceLevels defaults to a one-shot HTTP snapshot source. Set
    # PRICE_LEVELS_URL to point snapshots and quotes at a specific endpoint.
    url = os.getenv("PRICE_LEVELS_URL")
    prices = PriceLevels(PriceLevelsRpcSource(url)) if url else PriceLevels()

    # 1. Full snapshot: every pAMM's order book.
    snapshot = await prices.get_price_levels()
    block = snapshot.block_number if snapshot.block_number is not None else "?"
    print(f"snapshot @ block {block} — {len(snapshot.pamms)} pAMM(s)")

    # Pick the first pAMM/pair that carries an order book to display.
    chosen = next(
        (
            (entry.pamm, pair)
            for entry in snapshot.pamms
            for pair in entry.pairs
            if pair.order_book
        ),
        None,
    )
    if chosen is None:
        raise SystemExit("snapshot carried no order books")
    pamm, pair = chosen

    print(
        f"\n{_venue_name(pamm)} order book: "
        f"{_token_info(pair.token_in)[0]} -> {_token_info(pair.token_out)[0]}"
    )
    for level in pair.order_book:
        print(
            f"  {_fmt(level.amount_in, pair.token_in)} -> "
            f"{_fmt(level.amount_out, pair.token_out)}  [{level.variant}]"
        )

    # 2. Quote helpers, for that pair and a mid-ladder size. `get_quote` returns
    # the best across all pAMMs; `get_quote_venue` pins to one.
    size = pair.order_book[len(pair.order_book) // 2].amount_in

    best = await prices.get_quote(pair.token_in, pair.token_out, size)
    print(
        f"\nbest quote:  {_fmt(best.amount_in, best.token_in)} -> "
        f"{_fmt(best.amount_out, best.token_out)} via {_venue_name(best.pamm)}"
    )

    pinned = await prices.get_quote_venue(pamm, pair.token_in, pair.token_out, size)
    print(
        f"{_venue_name(pamm)} quote: {_fmt(pinned.amount_in, pinned.token_in)} -> "
        f"{_fmt(pinned.amount_out, pinned.token_out)}"
    )

    await prices.close()


if __name__ == "__main__":
    asyncio.run(main())

File: sdk/typescript/examples/price-levels.ts
/**
 * pAMM price levels: Titan's pre-quoted prices for a range of trade sizes,
 * grouped per pAMM. Served from Titan's latest snapshot over HTTP (no node RPC,
 * no key), unlike `router.quote`, which runs an on-chain `eth_call`.
 *
 * Build the SDK first, then run (Node >= 22.18 runs TypeScript directly):
 *
 *   pnpm build
 *   node examples/price-levels.ts
 *
 * Set PRICE_LEVELS_URL to use a specific HTTP endpoint.
 */
import { PriceLevels, PriceLevelsRpcSource } from "propamm/prices";
import type { PairPriceLevels } from "propamm/prices";
import { PAMMS } from "propamm/common/pamms";
import { formatUnits } from "propamm/common/helpers";
import type { Address } from "propamm";

// Known tokens, for readable amounts. Unknown tokens print raw (0 decimals).
const TOKENS: Record<string, { symbol: string; decimals: number }> = {
  "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48": { symbol: "USDC", decimals: 6 },
  "0xdac17f958d2ee523a2206206994597c13d831ec7": { symbol: "USDT", decimals: 6 },
  "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2": { symbol: "WETH", decimals: 18 },
  "0x2260fac5e5542a773aa44fbcfedf7c193bc2c599": { symbol: "WBTC", decimals: 8 },
};
const tokenInfo = (token: Address) => TOKENS[token.toLowerCase()] ?? { symbol: token, decimals: 0 };
const fmt = (amount: bigint, token: Address) => {
  const { symbol, decimals } = tokenInfo(token);
  return `${formatUnits(amount, decimals)} ${symbol}`;
};

// Map venue addresses back to readable pAMM names.
const venueNames: Record<string, string> = {};
for (const [name, address] of Object.entries(PAMMS)) {
  venueNames[address.toLowerCase()] = name;
}
const venueName = (address: Address) => venueNames[address.toLowerCase()] ?? address;

// PriceLevels defaults to a one-shot HTTP snapshot source. Set PRICE_LEVELS_URL
// to point snapshots and quotes at a specific endpoint.
const url = process.env.PRICE_LEVELS_URL;
const prices = new PriceLevels(url ? { source: new PriceLevelsRpcSource({ url }) } : {});

// 1. Full snapshot: every pAMM's order book.
const snapshot = await prices.getPriceLevels();
if (!snapshot) throw new Error("no price-levels snapshot available");
console.log(`snapshot @ block ${snapshot.blockNumber ?? "?"} — ${snapshot.pamms.length} pAMM(s)`);

// Pick the first pAMM/pair that actually carries an order book to display.
let chosen: { pamm: Address; pair: PairPriceLevels } | undefined;
for (const entry of snapshot.pamms) {
  const pair = entry.pairs.find((p) => p.orderBook.length > 0);
  if (pair) {
    chosen = { pamm: entry.pamm, pair };
    break;
  }
}
if (!chosen) throw new Error("snapshot carried no order books");
const { pamm, pair } = chosen;

console.log(
  `\n${venueName(pamm)} order book: ${tokenInfo(pair.tokenIn).symbol} -> ${tokenInfo(pair.tokenOut).symbol}`,
);
for (const level of pair.orderBook) {
  console.log(
    `  ${fmt(level.amountIn, pair.tokenIn)} -> ${fmt(level.amountOut, pair.tokenOut)}  [${level.variant}]`,
  );
}

// 2. Quote helpers, for that pair and a mid-ladder size. `getQuote` returns the
// best across all pAMMs; `getQuoteVenue` pins to one. Both skip the eth_call.
const size = pair.orderBook[Math.floor(pair.orderBook.length / 2)].amountIn;

const best = await prices.getQuote(pair.tokenIn, pair.tokenOut, size);
console.log(
  `\nbest quote:  ${fmt(best.amountIn, best.tokenIn)} -> ${fmt(best.amountOut, best.tokenOut)} via ${venueName(best.pamm)}`,
);

const pinned = await prices.getQuoteVenue(pamm, pair.tokenIn, pair.tokenOut, size);
console.log(
  `${venueName(pamm)} quote: ${fmt(pinned.amountIn, pinned.tokenIn)} -> ${fmt(pinned.amountOut, pinned.tokenOut)}`,
);

prices.close(); // close the stream socket so the process exits promptly
